# TASK-1 : Record Matching

### Link to the data file:
 [Record Matching Data File](https://docs.google.com/spreadsheets/d/1g99QKwm2V-_JMXtup7Nk081axpxRYbBx/edit#gid=1325724825)

### Description
- There are 2 datasets present in the file. Data 1 and Data 2
- The primary key for both data1 and data2 is Order Id + Product ID combination (i.e. the individual datasets do not have any duplicate on this combination)

### Provide solution and approach for the following:
1. How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2 (Specify the number of records missing in your answer)
2. How to identify the Records (Order ID + Product ID combination) missing in data1 but present in data2 (Specify the number of records missing in your answer)
3. Find the Sum of the total Qty of Records missing in data1 but present in data2
4. Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2

### Allowed Tools:
- Excel
- R
- Python
- SQL


### Importing necessary libraries and loading the data from excel file

In [3]:
import pandas as pd
import numpy as np

In [5]:
#Importing the data files
df_data1 = pd.read_excel('Records Matching Task.xlsx' , sheet_name = 'data1')
df_data2 = pd.read_excel('Records Matching Task.xlsx' , sheet_name = 'data2')

In [6]:
df_data1.head()

Order ID       Product ID  Qty
0  CA-2014-100006  TEC-PH-10002075    3
1  CA-2014-100090  FUR-TA-10003715    3
2  CA-2014-100090  OFF-BI-10001597    6
3  CA-2014-100293  OFF-PA-10000176    6
4  CA-2014-100328  OFF-BI-10000343    1

In [7]:
df_data2.head()

Order ID       Product ID  Qty
0  CA-2014-100090  FUR-TA-10003715    3
1  CA-2014-100090  OFF-BI-10001597    6
2  CA-2014-100293  OFF-PA-10000176    6
3  CA-2014-100328  OFF-BI-10000343    1
4  CA-2014-100363  OFF-FA-10000611    2

### Exploratory Data Analysis

In [12]:
df_data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9478 entries, 0 to 9477
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Order ID    9478 non-null   object
 1   Product ID  9478 non-null   object
 2   Qty         9478 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 222.3+ KB


In [13]:
df_data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9479 entries, 0 to 9478
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Order ID    9479 non-null   object
 1   Product ID  9479 non-null   object
 2   Qty         9479 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 222.3+ KB


In [14]:
df_data1.describe()

Qty
count  9478.000000
mean      3.789513
std       2.231866
min       1.000000
25%       2.000000
50%       3.000000
75%       5.000000
max      16.000000

In [15]:
df_data2.describe()

Qty
count  9479.000000
mean      3.797763
std       2.236052
min       1.000000
25%       2.000000
50%       3.000000
75%       5.000000
max      16.000000

In [10]:
df_data1.isnull().sum()

Order ID      0
Product ID    0
Qty           0
dtype: int64

In [11]:
df_data2.isnull().sum()

Order ID      0
Product ID    0
Qty           0
dtype: int64

### Creating the 'primary_key' column by concatenating the 'Order ID' and the 'Product ID'

In [16]:
df_data1["primary_key1"] = df_data1["Order ID"]+df_data1["Product ID"]
df_data2["primary_key2"] = df_data2["Order ID"]+df_data2["Product ID"]

In [17]:
df_data1.head()

Order ID       Product ID  Qty                   primary_key1
0  CA-2014-100006  TEC-PH-10002075    3  CA-2014-100006TEC-PH-10002075
1  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090FUR-TA-10003715
2  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090OFF-BI-10001597
3  CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293OFF-PA-10000176
4  CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328OFF-BI-10000343

In [18]:
df_data2.head()

Order ID       Product ID  Qty                   primary_key2
0  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090FUR-TA-10003715
1  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090OFF-BI-10001597
2  CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293OFF-PA-10000176
3  CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328OFF-BI-10000343
4  CA-2014-100363  OFF-FA-10000611    2  CA-2014-100363OFF-FA-10000611

### Merging the datasets data1 and data2 using primary-keys

In [19]:
df = df_data1.merge(df_data2 , how = 'outer' , left_on = ['primary_key1'] , right_on = ['primary_key2'])
df

Order ID_x     Product ID_x  Qty_x                   primary_key1  \
0     CA-2014-100006  TEC-PH-10002075    3.0  CA-2014-100006TEC-PH-10002075   
1     CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090FUR-TA-10003715   
2     CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090OFF-BI-10001597   
3     CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293OFF-PA-10000176   
4     CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328OFF-BI-10000343   
...              ...              ...    ...                            ...   
9981             NaN              NaN    NaN                            NaN   
9982             NaN              NaN    NaN                            NaN   
9983             NaN              NaN    NaN                            NaN   
9984             NaN              NaN    NaN                            NaN   
9985             NaN              NaN    NaN                            NaN   

          Order ID_y     Product ID_y  Qty_y                   primary_key2  
0                NaN              NaN    NaN                            NaN  
1     CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090FUR-TA-10003715  
2     CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090OFF-BI-10001597  
3     CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293OFF-PA-10000176  
4     CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328OFF-BI-10000343  
...              ...              ...    ...                            ...  
9981  US-2017-165344  OFF-BI-10003196   10.0  US-2017-165344OFF-BI-10003196  
9982  US-2017-165358  TEC-CO-10001943    5.0  US-2017-165358TEC-CO-10001943  
9983  US-2017-167920  OFF-AP-10000159    5.0  US-2017-167920OFF-AP-10000159  
9984  US-2017-169502  OFF-AP-10001947    5.0  US-2017-169502OFF-AP-10001947  
9985  US-2017-169551  FUR-BO-10001519    3.0  US-2017-169551FUR-BO-10001519  

[9986 rows x 8 columns]

In [20]:
df.isnull().sum()

Order ID_x      508
Product ID_x    508
Qty_x           508
primary_key1    508
Order ID_y      507
Product ID_y    507
Qty_y           507
primary_key2    507
dtype: int64

In [23]:
df.describe()

Qty_x        Qty_y
count  9478.000000  9479.000000
mean      3.789513     3.797763
std       2.231866     2.236052
min       1.000000     1.000000
25%       2.000000     2.000000
50%       3.000000     3.000000
75%       5.000000     5.000000
max      16.000000    16.000000

## Q1 : How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2

In [33]:
q1 = df[df['primary_key2'].isnull()]
q1

Order ID_x     Product ID_x  Qty_x                   primary_key1  \
0     CA-2014-100006  TEC-PH-10002075    3.0  CA-2014-100006TEC-PH-10002075   
10    CA-2014-100678  OFF-EN-10000056    3.0  CA-2014-100678OFF-EN-10000056   
19    CA-2014-100895  OFF-AR-10004511    2.0  CA-2014-100895OFF-AR-10004511   
35    CA-2014-101560  OFF-BI-10000309    3.0  CA-2014-101560OFF-BI-10000309   
61    CA-2014-102673  OFF-LA-10001771   12.0  CA-2014-102673OFF-LA-10001771   
...              ...              ...    ...                            ...   
9390  US-2017-160836  OFF-AP-10001626    2.0  US-2017-160836OFF-AP-10001626   
9403  US-2017-162558  FUR-FU-10002364    2.0  US-2017-162558FUR-FU-10002364   
9420  US-2017-163657  OFF-BI-10000138    5.0  US-2017-163657OFF-BI-10000138   
9427  US-2017-164056  FUR-TA-10001307    5.0  US-2017-164056FUR-TA-10001307   
9435  US-2017-165456  FUR-CH-10003981    6.0  US-2017-165456FUR-CH-10003981   

     Order ID_y Product ID_y  Qty_y primary_key2  
0           NaN          NaN    NaN          NaN  
10          NaN          NaN    NaN          NaN  
19          NaN          NaN    NaN          NaN  
35          NaN          NaN    NaN          NaN  
61          NaN          NaN    NaN          NaN  
...         ...          ...    ...          ...  
9390        NaN          NaN    NaN          NaN  
9403        NaN          NaN    NaN          NaN  
9420        NaN          NaN    NaN          NaN  
9427        NaN          NaN    NaN          NaN  
9435        NaN          NaN    NaN          NaN  

[507 rows x 8 columns]

In [34]:
q1.shape[0]

507

**Answer : 507**

## Q2 : How to identify the Records (Order ID + Product ID combination) missing in data1 but present in data2

In [35]:
q2 = df[df['primary_key1'].isnull()]
q2

Order ID_x Product ID_x  Qty_x primary_key1      Order ID_y  \
9478        NaN          NaN    NaN          NaN  CA-2014-100706   
9479        NaN          NaN    NaN          NaN  CA-2014-100762   
9480        NaN          NaN    NaN          NaN  CA-2014-101427   
9481        NaN          NaN    NaN          NaN  CA-2014-102652   
9482        NaN          NaN    NaN          NaN  CA-2014-102869   
...         ...          ...    ...          ...             ...   
9981        NaN          NaN    NaN          NaN  US-2017-165344   
9982        NaN          NaN    NaN          NaN  US-2017-165358   
9983        NaN          NaN    NaN          NaN  US-2017-167920   
9984        NaN          NaN    NaN          NaN  US-2017-169502   
9985        NaN          NaN    NaN          NaN  US-2017-169551   

         Product ID_y  Qty_y                   primary_key2  
9478  TEC-AC-10001314    2.0  CA-2014-100706TEC-AC-10001314  
9479  OFF-PA-10001815    3.0  CA-2014-100762OFF-PA-10001815  
9480  OFF-AR-10002257    3.0  CA-2014-101427OFF-AR-10002257  
9481  FUR-FU-10001918    7.0  CA-2014-102652FUR-FU-10001918  
9482  OFF-PA-10000788    3.0  CA-2014-102869OFF-PA-10000788  
...               ...    ...                            ...  
9981  OFF-BI-10003196   10.0  US-2017-165344OFF-BI-10003196  
9982  TEC-CO-10001943    5.0  US-2017-165358TEC-CO-10001943  
9983  OFF-AP-10000159    5.0  US-2017-167920OFF-AP-10000159  
9984  OFF-AP-10001947    5.0  US-2017-169502OFF-AP-10001947  
9985  FUR-BO-10001519    3.0  US-2017-169551FUR-BO-10001519  

[508 rows x 8 columns]

In [36]:
q2.shape[0]

508

**Answer : 508**

## Q3 : Find the Sum of the total Qty of Records missing in data1 but present in data2

In [37]:
q2['Qty_y'].sum()

1956.0

**Answer : 1956.0**

## Q4 : Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2

In [32]:
df.shape[0]

9986

**Answer : 9986**